File: quickmeal-frontend/src/services/orderService.ts
// src/services/orderService.ts
import api from "./api";
import type { OrderResponseDTO } from "@/types";

// Interface cho Spring Data Page
export interface PageResponse<T> {
    content: T[];
    totalPages: number;
    totalElements: number;
    size: number;
    number: number;
}

export interface OrderItemRequest {
    productId: number;
    quantity: number;
}

export interface OrderRequestDTO {
    userName: string;
    address: string;
    phone: string;
    note: string;
    items: OrderItemRequest[];
}

export const orderService = {
    createOrder: async (data: OrderRequestDTO): Promise<OrderResponseDTO> => {
        const response = await api.post("/orders/checkout", data);
        return response.data;
    },

    getOrders: async (
        page = 0,
        size = 10,
        keyword = "",
        status = "ALL" // Mặc định lấy tất cả
    ): Promise<PageResponse<OrderResponseDTO>> => {
        const params: any = { page, size, keyword, sort: "createdAt,desc" };

        // Nếu khác ALL thì mới gửi status lên filter
        if (status !== "ALL") {
            params.status = status;
        }

        const res = await api.get("/orders", { params });
        return res.data;
    },

    updateStatus: async (id: number, status: string): Promise<OrderResponseDTO> => {
        const response = await api.patch(`/orders/${id}/status?status=${status}`);
        return response.data;
    },
    getOrderById: async (id: number): Promise<OrderResponseDTO> => {
        const response = await api.get(`/orders/${id}`);
        return response.data;
    },
    getMyOrders: async (
        userName: string,
        page: number = 0,
        size: number = 10,
        status: string = "ALL"
    ): Promise<PageResponse<OrderResponseDTO>> => {
        const params: any = { userName, page, size, sort: "createdAt,desc" };
        if (status !== "ALL") {
            params.status = status;
        }
        const res = await api.get(`/orders/my-orders`, { params });
        return res.data;
    },
};

File: quickmeal-backend/src/main/java/com/quickmeal/backend/service/OrderService.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.quickmeal.backend.service;

import com.quickmeal.backend.constant.OrderStatus;
import com.quickmeal.backend.dto.order.OrderRequestDTO;
import com.quickmeal.backend.dto.order.OrderResponseDTO;
import com.quickmeal.backend.entity.*;
import com.quickmeal.backend.repo.*;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.util.StringUtils;

/**
 *
 * @author <a href="https://www.facebook.com/khanhdepzai.pro/">KhanhDzai</a>
 */
@Service
@RequiredArgsConstructor
public class OrderService {

    private final OrderRepository orderRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;

    @Transactional(readOnly = true)
    public Page<OrderResponseDTO> getAllOrders(String keyword, String statusStr, Pageable pageable) {
        String searchKeyword = StringUtils.hasText(keyword) ? keyword : "";

        // Chuyển từ String sang Enum, nếu là "ALL" hoặc null thì để null để Repo không filter
        OrderStatus status = null;
        if (StringUtils.hasText(statusStr) && !"ALL".equalsIgnoreCase(statusStr)) {
            try {
                status = OrderStatus.valueOf(statusStr);
            } catch (IllegalArgumentException e) {
                // Log warning nếu status gửi lên không hợp lệ
                status = null;
            }
        }

        return orderRepository.searchOrdersAdvanced(searchKeyword, status, pageable)
                .map(this::convertToDTO);
    }

    private OrderResponseDTO convertToDTO(OrderEntity entity) {
        return OrderResponseDTO.builder()
                .id(entity.getId())
                .customerName(entity.getUser().getFullName())
                .userName(entity.getUser().getUserName())
                .phone(entity.getPhone())
                .address(entity.getAddress())
                .note(entity.getNote())
                .totalPrice(entity.getTotalPrice())
                .status(entity.getStatus())
                .createdAt(entity.getCreatedAt())
                .items(entity.getItems().stream()
                        .map(item -> OrderResponseDTO.OrderItemDTO.builder()
                        .productName(item.getProduct().getName())
                        .quantity(item.getQuantity())
                        .price(item.getPrice())
                        .build())
                        .collect(Collectors.toList()))
                .build();
    }

    @Transactional
    public OrderResponseDTO createOrder(OrderRequestDTO dto) {
        UserEntity user = userRepository.findByUserName(dto.getUserName())
                .orElseThrow(() -> new RuntimeException("User không tồn tại: " + dto.getUserName()));

        OrderEntity order = OrderEntity.builder()
                .user(user)
                .address(dto.getAddress())
                .phone(dto.getPhone())
                .note(dto.getNote())
                .status(OrderStatus.PENDING_ACCEPTANCE)
                .totalPrice(0.0)
                .build();

        List<OrderItemEntity> items = dto.getItems().stream().map(itemDto -> {
            ProductEntity product = productRepository.findById(itemDto.getProductId())
                    .orElseThrow(() -> new RuntimeException("Sản phẩm ID " + itemDto.getProductId() + " không tồn tại"));

            return OrderItemEntity.builder()
                    .order(order)
                    .product(product)
                    .quantity(itemDto.getQuantity())
                    .price(product.getPrice())
                    .build();
        }).toList();

        double total = items.stream().mapToDouble(i -> i.getPrice() * i.getQuantity()).sum();
        order.setTotalPrice(total);
        order.setItems(items);

        OrderEntity savedOrder = orderRepository.save(order);
        return convertToDTO(savedOrder);
    }

    @Transactional
    public OrderResponseDTO updateStatus(Long orderId, OrderStatus newStatus) {
        OrderEntity order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Không tìm thấy đơn hàng"));
        order.setStatus(newStatus);
        order.setUpdatedAt(LocalDateTime.now());

        OrderEntity savedOrder = orderRepository.save(order);
        return convertToDTO(savedOrder);
    }

    @Transactional(readOnly = true)
    public Page<OrderResponseDTO> getOrdersByCustomer(String userName, String statusStr, Pageable pageable) {
        if (!userRepository.existsByUserName(userName)) {
            throw new RuntimeException("Người dùng không tồn tại: " + userName);
        }

        // Logic chuyển đổi Status chuỗi sang Enum
        OrderStatus status = null;
        if (StringUtils.hasText(statusStr) && !"ALL".equalsIgnoreCase(statusStr)) {
            try {
                status = OrderStatus.valueOf(statusStr);
            } catch (IllegalArgumentException e) {
                status = null; // Nếu sai enum thì mặc định coi như lấy ALL
            }
        }

        return orderRepository.findByUserNameAndStatus(userName, status, pageable)
                .map(this::convertToDTO);
    }
}
